src/main/kotlin/com/hypto/iam/server/service/PasscodeService.kt

```diff
@@ -0,0 +1,67 @@
+package com.hypto.iam.server.service
+
+import com.hypto.iam.server.configs.AppConfig
+import com.hypto.iam.server.db.repositories.PasscodeRepo
+import com.hypto.iam.server.exceptions.PasscodeLimitExceededException
+import com.hypto.iam.server.models.BaseSuccessResponse
+import com.hypto.iam.server.models.VerifyEmailRequest.Purpose
+import com.hypto.iam.server.utils.ApplicationIdUtil
+import java.time.LocalDateTime
+import java.util.Base64
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.inject
+import software.amazon.awssdk.services.ses.SesClient
+import software.amazon.awssdk.services.ses.model.Destination
+import software.amazon.awssdk.services.ses.model.SendTemplatedEmailRequest
+
+class PasscodeServiceImpl : KoinComponent, PasscodeService {
+    private val sesClient: SesClient by inject()
+    private val appConfig: AppConfig by inject()
+    private val idGenerator: ApplicationIdUtil.Generator by inject()
+    private val passcodeRepo: PasscodeRepo by inject()
+
+    override suspend fun verifyEmail(
+        email: String,
+        purpose: Purpose,
+        organizationId: String?
+    ): BaseSuccessResponse {
+        if (passcodeRepo.getValidPasscodeCount(email, purpose) >= appConfig.app.passcodeCountLimit) {
+            throw PasscodeLimitExceededException(
+                "You can only send ${appConfig.app.passcodeCountLimit} passcodes per day"
+            )
+        }
+        val validUntil = LocalDateTime.now().plusSeconds(appConfig.app.passcodeValiditySeconds)
+        val passcode = passcodeRepo.createPasscode(
+            idGenerator.passcodeId(),
```

Comment:
Lets ensure that "idGenerator" generated passcode will not generate the master key at anytime. We can ensure this by one of the below option
1. Having different length requirement for master key vs passcode
2. Have different character set when generating passcode vs master key.

Reply:
https://github.com/hwslabs/iam/blob/1bf37f1e44b7659e573fec24809f049fca932e29/src/main/kotlin/com/hypto/iam/server/utils/MasterKeyUtil.kt#L16-L20
https://github.com/hwslabs/iam/blob/1bf37f1e44b7659e573fec24809f049fca932e29/src/main/resources/db/migration/V1__init_iam_database.sql#L119-L120

Master Key generation is done using a script and our id for that is an uuid. So i think it will not collide with our passcode.

Reply:
By master key, i meant the x-api-key which is provided as part of appConfig "appConfig.app.secretKey" in the "hypto-iam-root-auth". We should ensure that at no point, random generator should generate this key.